### ISSUE TITLE
Versions/upgrades

### ISSUE THREAD

**Issue by shermp** [pgaskin/NickelHook#1]
I noticed your little note regarding a versioning mechanism. Is it too early to discuss this? 😀 

One option is to do something similar to what I prototyped in NDB. Another potential option might be some sort of external version file(s).

Ideally, a mod should be able to be included in a KoboRoot.tgz without fear of clobbering a newer version that is already installed.

**Comment by pgaskin** [comment]
The most obvious solution to this would be to have an exported function which returns the current version (rather than directly accessing the NH struct, as that could change). But, from there, you run into the issue of NH running in the library's constructor (this could be fixed by including it in an ELF section, but I'd rather not work with those if I have another choice).

Another solution would be to store the version in the filename, and if a version is present in the current file's filename, it compares itself to the others and removes itself if it's not the latest. And, if one doesn't have a version, it would take priority over the others.

On the other hand, users should be able to downgrade if they want.

**Comment by redphx** [comment]
Can we use plugin's JSON metadata file? https://runebook.dev/en/docs/qt/qpluginloader/metadata

**Comment by pgaskin** [comment]
No, since NickelHook doesn't use plugins as anything but a way to get loaded (they don't load/unload cleanly).
